libmobilecoin/src/transaction.rs

```diff
@@ -18,22 +19,44 @@ use mc_util_ffi::*;
 /* ==== TxOut ==== */
 
 #[repr(C)]
-pub struct McTxOutAmount<'a> {
+pub struct McTxOutAmount {
     /// 32-byte `CompressedCommitment`
-    commitment: FfiRefPtr<'a, McBuffer<'a>>,
     masked_value: u64,
 }
 
-impl<'a> TryFromFfi<&McTxOutAmount<'a>> for Amount {
-    type Error = LibMcError;
+/// # Preconditions
+///
+/// * `view_private_key` - must be a valid 32-byte Ristretto-format scalar.
+#[no_mangle]
+pub extern "C" fn mc_tx_out_reconstruct_commitment(
+    tx_out_amount: FfiRefPtr<McTxOutAmount>,
+    tx_out_public_key: FfiRefPtr<McBuffer>,
+    view_private_key: FfiRefPtr<McBuffer>,
+    out_tx_out_commitment: FfiMutPtr<McMutableBuffer>,
+    out_error: FfiOptMutPtr<FfiOptOwnedPtr<McError>>
+) -> bool {
+    ffi_boundary_with_error(out_error, || {
+        let view_private_key = RistrettoPrivate::try_from_ffi(&view_private_key)
+            .expect("view_private_key is not a valid RistrettoPrivate");
 
-    fn try_from_ffi(src: &McTxOutAmount<'a>) -> Result<Self, Self::Error> {
-        let commitment = CompressedCommitment::try_from_ffi(&src.commitment)?;
-        Ok(Amount {
-            commitment,
-            masked_value: src.masked_value,
-        })
-    }
+        let tx_out_public_key = RistrettoPublic::try_from_ffi(&tx_out_public_key)?;
+
+        let shared_secret = get_tx_out_shared_secret(&view_private_key, &tx_out_public_key);
+        
+        let value =
+            (tx_out_amount.masked_value as u64) ^ get_value_mask(&shared_secret);
+
+        let amount: Amount = Amount::new(value, &shared_secret)
+            .expect("could not create amount object");
```

Comment:
I think it's better not to use `.expect` here, because if the error happens, the app will crash. I think usually we are using `?` to pass the error back, ultimately as a human readable string error message?

Reply:
good idea, Ill make that change.

As a side-note: how can I know what Error types the function may throw/return ? I see those enumerated in the documentation comments and Im wondering if there is an automated way for me to figure it out for a given function.

Reply:
I think I'm just reading the code and looking at what appear to be the failure paths, so it's literally like, if there is `return Err(...);` or `?` which is a shorter syntax for `return Err(...);`, then those are the paths where it can return an error, and otherwise, it would only be like, if we have `return foo(); ` and `foo` returns `Result<..., ...>` and so could also be an error.

Reply:
Usually like, if a function return `Result<Foo, Error>` then `Error` will be an enum and you can look at the cases -- in this code I think all the errors get stringified before going across FFI? not sure, would have to check

Reply:
[Plug for CLion+Rust, with it's inline type hints, goes here]

In `Rust`, there are no exceptions, and `Result<T, E>` is actually just a type-parameterized enum:

```
pub enum Result<T: Sized, E: Sized> {
    Ok(T),
    Err(E),
}
```

So when you do `return Err(SomeError::ErrCode);`, you're really creating an concrete instance of that `Result` enum, and returning it by value---there's zero magic around returning errors in the usual case, you're just creating a new `Result::Err` enum value, which contains the `SomeError` value inside it, and returning that enum-in-an-enum to the caller, by value. LLVM is really good at optimizing this case, so it's not typically a real hit to performance.

And, as Chris mentioned, the `?` is syntactic sugar that resolves:

```
fn inner_foo() -> Result<u64, InnerError> { /* ... */ }

fn outer_foo() -> Result<i128, OuterError> {
    let value = do_foo()?;
    /* ... */
    value as i128
}
```
into this:
```
fn inner_foo() -> Result<u64, InnerError> { /* ... */ }

fn outer_foo() -> Result<i128, OuterError> {
    let value = {
        let result = do_foo();
        match result {
            Ok(value) => value,
            Err(error) => return Err(OuterError::from(error)), // early return
        }
    };
    /* ... */
    value as i128
}
```